common4j/src/testFixtures/java/com/microsoft/identity/http/HttpRequestRewriter.java

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.http;
+
+import com.microsoft.identity.common.java.net.HttpClient;
+import com.microsoft.identity.common.java.net.HttpResponse;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.Map;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
+import lombok.AllArgsConstructor;
+import lombok.NonNull;
+
+/**
+ * An {@link HttpRequestInterceptor} that allows to re-write the http request and then proceeds
+ * with the modified request.
+ */
+@AllArgsConstructor
+public abstract class HttpRequestRewriter implements HttpRequestInterceptor {
```

Comment:
I can see how the other ones are being used (examiner = check request content, rewriter = modify response), but i'm not sure what's the purpose of this one.

Reply:
That said, if I were to write test, i would probably need both examier and rewriter in one class (check if request is valid, and mock the response).

Reply:
Intercept the request as it is about to go in the wire, and then modify something in the request and then send it on the wire.

For example, one may choose to remove a particular header from request headers and then send it forward.

Reply:
Yes, I think what we need is the ability to chain multiple interceptors together to be able to examine, re-write request as well as rewrite response; all for one request, however, the existing model in the code didn't make it so easy to implement that. What we've done here is provided an improvement to the existing model and an opportunity to ourselves to take it and improve it further.